### ISSUE TITLE
Control of the ClearScript Engine

### ISSUE THREAD

**Issue by Rison-Hub** [microsoft/ClearScript#574]
In a script executed by the ClearScript engine, there's a for loop performing a heartbeat-like function. I want to achieve the following:

Send a pause command to the engine to pause the entire engine and halt this for loop.
Send a start command to the engine to restart the engine and resume this for loop.
Send a stop command to the engine to terminate the currently executing script.

**Comment by ClearScriptLib** [comment]
Hi @Rison-Hub,

The only way to achieve that is to have the host and script implement a play/pause/exit protocol of some sort.

For example, the host could expose a property that the script periodically checks to determine whether to continue, await a continuation signal, or exit.

Good luck!

**Comment by Rison-Hub** [comment]
Can you provide me with a simple demo for reference?

**Comment by ClearScriptLib** [comment]
Hi @Rison-Hub,

>Can you provide me with a simple demo for reference?

Sure. Here's a class that provides basic pause/resume/terminate support:

``` C#
public class ScriptControl {
    private readonly ManualResetEventSlim _notPaused = new(true);
    private bool _terminationRequested;
    public void Pause() => _notPaused.Reset();
    public void Resume() => _notPaused.Set();
    public void Terminate() => _terminationRequested = true;
    public bool CanContinue() {
        if (_terminationRequested) return false;
        _notPaused.Wait();
        return true;
    }
}
```

To use it, let's kick off a script execution thread:

``` C#
var scriptControl = new ScriptControl();
var scriptThread = new Thread(() => {
    using var engine = new V8ScriptEngine();
    engine.AddHostType(typeof(Console));
    engine.Script.canContinue = new Func<bool>(scriptControl.CanContinue);
    engine.Script.sleep = new Action<int>(Thread.Sleep);
    engine.Execute(@"
        while (true) {
            if (!canContinue()) break;
            Console.WriteLine('Script running...');
            sleep(500);
        }
    ");
});
scriptThread.Start();
```

We can now control script execution as follows:

``` C#
Thread.Sleep(2000);
scriptControl.Pause();
Console.WriteLine("Script paused for 5 seconds.");
Thread.Sleep(5000);
scriptControl.Resume();
Thread.Sleep(2000);
scriptControl.Terminate();
scriptThread.Join();
Console.WriteLine("Script terminated.");
```

This is just a simple example, but hopefully it gives you an idea of what's possible.

Cheers!

**Comment by Rison-Hub** [comment]
thank you

**Comment by Rison-Hub** [comment]
No matter where the JavaScript code is written, as long as the "Pause" button is clicked, the entire script execution can be paused — including non-loop code such as a = 5;, c = a + b;, etc.

After the loop, there’s also:

`while (true) {
            if (!canContinue()) break;
            Console.WriteLine('Script running...');
            sleep(500);
        }
var a, b, c;
a = 5;
b = 6;
c = a + b;
Console.WriteLine('C=' + c);
zp.Wait(500);
Console.WriteLine('Script End...');
`
When the execution reaches c = a + b;, I want to press "Pause" to suspend the thread so that it doesn’t continue executing.

Then, when I click "Continue" again, it should unblock and the program should resume execution, printing:

mathematica
C=11
Script End...
How can I implement this?

**Comment by ClearScriptLib** [comment]
>When the execution reaches c = a + b;, I want to press "Pause" to suspend the thread so that it doesn’t continue executing.

Please clarify. There's no way to reliably press "Pause" exactly when script execution reaches a specific line. Or are you asking about the ability to pause script execution programmatically – that is, by calling something from script code?

**Comment by Rison-Hub** [comment]
> > When the execution reaches c = a + b;, I want to press "Pause" to suspend the thread so that it doesn’t continue executing.
> 
> Please clarify. There's no way to reliably press "Pause" exactly when script execution reaches a specific line. Or are you asking about the ability to pause script execution programmatically – that is, by calling something from script code?

I have a software with the following functional requirements: It can execute a JS script file and then print the script's output information within the software for easy viewing. If any issues occur, I can press pause at any time to halt the execution of the JS script, then review the output information to troubleshoot the problem. After that, I can press resume to continue running the script.
Can it pause/block execution at any arbitrary point in the script to implement the suspend functionality?

**Comment by ClearScriptLib** [comment]
>Can it pause/block execution at any arbitrary point in the script to implement the suspend functionality?

No. The only way is for the script and host to work together as shown above. V8 does not provide a way to suspend and resume script execution.

**Comment by Rison-Hub** [comment]
OK，thanks
